File: 0032_Longest_Valid_Parentheses/Longest_Valid_Parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        stack = [-1]     # base index for calculating lengths
        maxlen = 0

        for i, ch in enumerate(s):
            if ch == '(':
                stack.append(i)
            else:  # ch == ')'
                # pop the matching '(' index (if any)
                stack.pop()
                if not stack:
                    # no unmatched '(' to anchor a valid substring -> set new base
                    stack.append(i)
                else:
                    # length of current valid substring ending at i
                    maxlen = max(maxlen, i - stack[-1])

        return maxlen

File: 0032_Longest_Valid_Parentheses/Longest_Valid_Parentheses.cpp
class Solution {
public:
    int longestValidParentheses(string s) {

        // We store indices of characters of s in this vector stack.
        // Start with -1 as a "base boundary" to mark the position
        // before the beginning of any valid substring.
        vector<int> stack = {-1};

        int maxlen = 0;

        for (int i = 0; i < s.size(); i++) {

            // If we see an opening bracket, it may start a future pair.
            // So we store its index.
            if (s[i] == '(') {
                stack.push_back(i);
            }

            // If we see a closing bracket:
            else {

                // Try to match it with the most recent '('.
                // Pop the last stored index.
                stack.pop_back();

                // Case 1: stack becomes empty after popping —
                // This means the ')' had no matching '('.
                //
                // So this index becomes the new "boundary".
                // No valid substring can cross this point.
                if (stack.empty()) {
                    stack.push_back(i);
                }

                // Case 2: stack is NOT empty —
                // This means the current ')' has a matching '('.
                //
                // The length of the current valid substring is:
                // current index - index of the last unmatched bracket/boundary.
                else {
                    int currlen = i - stack.back();
                    maxlen = max(maxlen, currlen);
                }
            }
        }

        return maxlen;
    }
};
